### DISCUSSION TITLE
Does CloudFunctionInvokeFunctionOperator work with Gen2 GCP CloudFunctions ?

### DISCUSSION THREAD

**Discussion by murfee25** [apache/airflow#32688] [Q&A]
I'm trying to call a GCP Cloud function from a DAG, 

```
resp= CloudFunctionInvokeFunctionOperator(
        task_id = "invoke_cf",
        project_id = 'my-project-id',
        location = 'europe-west3',
        input_data = {"data": json.dumps({"key1":"key1 value", "key2": "testing" })},
        function_id = 'my-function-name'
    )
```

However I get 
```
HttpError 404 when requesting https://cloudfunctions.googleapis.com/v1/projects/my-project-id/locations/europe-west3/functions/my-function-name:call?alt=json

Function my-function-name  in region europe-west3 in project my-project-id does not exist
```

But the function definitely does exist in my project and in that region.

My question is does this operator and/or the hook `CloudFunctionsHook` work with Gen2 GCP CloudFunctions ?

I notice when I test the function in GCP Console it uses a url like : `https://europe-west3-my-project-id.cloudfunctions.net/my-function-name`

**Comment by potiuk** [comment]
Wild guess is that you have no permissions, but I have no idea you might try to see the history of changes for the relevant file and ping the author or people who made substantial changes there,  Airflow is created by > 2500 people - the relevant people might not notice you started a discussion, so if you did not get help so far and still need it, that might be the way forward to try to find and  ping those who contributed it.

**Comment by mooniker** [comment]
I'm experiencing the same issue. @murfee25 did you ever figure it out?

**Reply by murfee25** [reply]
@mooniker, 

I'm afraid I came to the conclusion it didn't support gen2 functions.
Perhaps there is a better and more elegant way to do it but I ended up writing a class to wrap calling my gen2 function. 
This class would get an id_token and call the function with url pattern 
`function_url = f'https://{location}-{gcp_project}.cloudfunctions.net/{function_name}'
`

I use composer-local-dev in my local docker which used `gcloud auth application-default login` to authenticate with GCP. But when running in Cloud Composer it required the `fetch_id_token` call.

```python 
import os
import google.auth
from google.auth.transport.requests import Request
import requests
import google.auth.transport.requests
from  google.oauth2.id_token import fetch_id_token

class GcpFunction:
    def __init__(self, function_url) -> None:
        self.function_url = function_url
    
    def call(self, payload):
        token = self._get_id_token()

        #Set up the headers with the ID token
        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'  # Set the content type as needed
        }

        try:
          resp = requests.post(self.function_url, json=payload, headers=headers)
          resp.raise_for_status()
        except Exception as ex:
            print(f"GcpFunction.call {self.function_url} failed. {ex}")
            raise

        try:
            return resp.json()
        except ValueError:
            # If the response is not valid JSON, return the response content as a string
            return resp.text

    def _get_id_token(self):
        credentials, _ = google.auth.default()
        token_req = Request()
        credentials.refresh(token_req)
        
        # If running locally, will find id_token on credentials object, otherwise need to call fetch_id_token.
        id_token = credentials.id_token if hasattr(credentials, 'id_token') else fetch_id_token(token_req, self.function_url)

        return id_token
        
```

Hope that helps.

**Comment by szxyks** [comment]
I'm having the same issue as the original poster when using "`CloudFunctionInvokeFunctionOperator`" to call a Cloud Functions second generation

**Comment by houtanb** [comment]
FYI the problem with v2 cloud functions has been flagged, even though the proposed solution is not very helpful if you have both v1 and v2 functions https://github.com/apache/airflow/issues/42146

However, you can just write the code to authenticate and call the function directly without relying on support from the GCP airflow library.

```python
@task
def cloud_function_task(endpoint):
    """Call cloud function represented by endpoint."""
    auth_req = google.auth.transport.requests.Request()
    id_token = google.oauth2.id_token.fetch_id_token(auth_req, endpoint)
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {id_token}",
    }

    response = requests.post(
        url=endpoint,
        headers=headers,
        data=None,
    )

    if response.status_code != 200:
        raise Exception(f"\nError invoking {endpoint}: {response.text}\n")

    return response.text
```

Your service account will need permissions to invoke the given cloud functions.

If running locally, you'll need to set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable to the credentials file.
